contribute/design/1.x/cstor/20200511-cstor-backupandrestore.md

```diff
@@ -0,0 +1,259 @@
+
+As mentioned [above](#openebs-veleroplugin) velero-plugin triggers different REST API calls based on the type of request from velero-server.
+
+### Steps to perform user stories
+
+- User can create backup using velero CLI
+  Ex: velero backup create <BACKUP_NAME> --include-namespaces=<NAME_SPACE> --snapshot-volumes –volume-snapshot-locations=<SNAPSHOT_LOCATION>
+- User can resotre backuped using velero CLI
+  Ex: velero restore create --from-backup <BACKUP_NAME> --restore-volumes=true
+
+### Low Level Design
+
+#### Work Flow
+
+##### Velero Server sends CreateSnapshot Request
+
+Velero server sends CreateSnapshot API with volumeID to create the snapshot. Velero Interface will `CreateSnapshot` API of the cStor backup/restore controller. cStor backup/restore controller is responsible for executing below steps:
+1. BackUp the relevant PVC object to the cloud provider.
+2. Execute the REST API(POST `/latest/backups`) of CVC-Operator (This object will include the IP Address of snapshot receiver/sender module) to create backup.
+    2.1 Create a snapshot using volume name and snapshot name(which will get during the CreateSnapshot request).
+    2.2 Find the Healthy cStorVolumeReplica if it doesn't exist return error.
+    2.3 Create CStorCompleteBackUp resources if it doesn’t exist(Intention of creating this resource is used for incremental backup purpose).
+    2.4 Create CStorBackUp resource by populating the current snapshot name and previous snapshot name(if exists from CStorCompleteBackUp) with Healthy CStorVolumeReplica pool UID.
+    2.5 Corresponding backup controllers exist in pool-manager will responsible for sending the snapshot data from pool to velero-plugin.         and velero-plugin will write this stream to cloud-provider.
+3. Call cloud interface API `UploadSnapshot` which will upload the backup to the cloud provider.
+4. This API will return the unique ‘snapshotID’ (volumeID + "-velero-bkp-" + backupName) to the velero server. This ‘snapshotID’ will be used to refer to the backup snapshot.
+
+##### Velero Server Sends DeleteSnapshot Request
+
+Velero server sends `DeleteSnapshot` API of Velero Interface to delete the backup/snapshot with argument `snapshotID`. This snapshotID is generated during the backup creation of this snapshot. Velero Interface will execute the DeleteSnapshot API of cStor backup/restore module. cStor backup/restore module is responsible for performing below steps:
+1. Delete the PVC object for this backup from the cloud provider.
+2. Execute REST API(`latest/backups/`) of the CVC-Operater to delete the resources created for this particular backup.
+    2.1 Delete the CStorCompleteBackup resources if it is not a complete backup or if that is the only backup exists for that volume(So next backup will be complete backup).
```

Comment:
We are deleting cstorcompletedbackup, if the given cstorbackup is the last backup of schedule or cstorcompletedbackup doesn't have any successful backup.

Reply:
Updated PTAL